agent/ebs/watcher_test.go

```diff
@@ -70,233 +70,318 @@ func newTestEBSWatcher(ctx context.Context, agentState dockerstate.TaskEngineSta
-	mockTaskEngine.EXPECT().GetDaemonTask(md.EbsCsiDriver).Return(nil).AnyTimes()
-	mockTaskEngine.EXPECT().GetDaemonManagers().Return(nil).AnyTimes()
-
-	mockCsiClient := mock_csiclient.NewMockCSIClient(mockCtrl)
-	mockCsiClient.EXPECT().NodeStageVolume(gomock.Any(),
-		taskresourcevolume.TestVolumeId,
-		gomock.Any(),
-		filepath.Join(hostMountDir, taskresourcevolume.TestSourceVolumeHostPath),
-		taskresourcevolume.TestFileSystem,
-		gomock.Any(),
-		gomock.Any(),
-		gomock.Any(),
-		gomock.Any(),
-		gomock.Any()).Return(nil).AnyTimes()
-
-	testAttachmentProperties := map[string]string{
-		apiebs.DeviceNameKey:           taskresourcevolume.TestDeviceName,
-		apiebs.VolumeIdKey:             taskresourcevolume.TestVolumeId,
-		apiebs.VolumeNameKey:           taskresourcevolume.TestVolumeName,
-		apiebs.SourceVolumeHostPathKey: taskresourcevolume.TestSourceVolumeHostPath,
-		apiebs.FileSystemKey:           taskresourcevolume.TestFileSystem,
-		apiebs.VolumeSizeGibKey:        taskresourcevolume.TestVolumeSizeGib,
-	}
-
-	expiresAt := time.Now().Add(time.Millisecond * testconst.WaitTimeoutMillis)
-	ebsAttachment := &apiebs.ResourceAttachment{
-		AttachmentInfo: attachment.AttachmentInfo{
-			TaskARN:              taskARN,
-			TaskClusterARN:       taskClusterARN,
-			ContainerInstanceARN: containerInstanceARN,
-			ExpiresAt:            expiresAt,
-			Status:               attachment.AttachmentNone,
-			AttachmentARN:        resourceAttachmentARN,
+func TestHandleEBSAttachment(t *testing.T) {
+	tcs := []struct {
+		name                          string
+		shouldTick                    bool
+		shouldHandle                  bool
+		setDaemonRunningExpectations  func(*gomock.Controller, *mock_engine.MockTaskEngine, *mock_daemonmanager.MockDaemonManager, chan statechange.Event)
+		setCSIClientExpectations      func(*gomock.Controller, *mock_csiclient.MockCSIClient)
+		setConfirmEBSVolumeIsAttached func(*gomock.Controller, *mock_ebs_discovery.MockEBSDiscovery)
+		expectedNumSavedAttachments   int
+		attachments                   []*apiebs.ResourceAttachment
+		numTicks                      int
+	}{
+		{
+			name:         "handle with new CSI driver task",
+			shouldTick:   true,
+			numTicks:     2,
+			shouldHandle: true,
+			setDaemonRunningExpectations: func(ctrl *gomock.Controller, mockTaskEngine *mock_engine.MockTaskEngine, mockDaemonManager *mock_daemonmanager.MockDaemonManager, stateChangeChan chan statechange.Event) {
+				daemonManagers := map[string]daemonmanager.DaemonManager{
+					md.EbsCsiDriver: mockDaemonManager,
+				}
+				gomock.InOrder(
+					mockTaskEngine.EXPECT().GetDaemonTask(md.EbsCsiDriver).Return(nil).Times(1),
+					mockTaskEngine.EXPECT().GetDaemonTask(md.EbsCsiDriver).Return(&task.Task{
+						Arn:               "arn:aws:ecs:us-east-1:012345678910:task/some-task-id",
+						KnownStatusUnsafe: status.TaskRunning,
+					}).Times(1),
+				)
+				mockTaskEngine.EXPECT().GetDaemonManagers().Return(daemonManagers).Times(1)
```

Comment:
nit: `Times(1)` is redundant as by default gomock sets the number of expected calls to 1.

Reply:
Good point, I was thinking more of a max amount of times it should be called. Perhaps I can use `MaxTimes()` as well.

Reply:
`.Times(1)` (and not having any `.TImes(1)` since 1 is the default) will make the mock check that the method was invoked one and only once.

Reply:
TIL